### DISCUSSION TITLE
Best approach for scheduling update checks (and installation) even if hostBundle is not running

### DISCUSSION THREAD

**Discussion by Davack** [sparkle-project/Sparkle#2169] [Q&A] [answered]
Hey everyone,
I've been using Sparkle for some time, and I love it!
There is one scenario where I don't find any formal solution within Sparkle.
I want all my clients to check for updates in a scheduled manner (let's say once a day) and will download & install updates if found. This way, my clients will be much more up-to-date when launching my app.

The solutions I came up with:
1. Create a launch daemon agent that uses Sparkle with `initWithHost: myAppHost` and check for updates every once in a while.
2. Sending XPC message to trigger Sparkle update.

I will be glad to both hear your ideas and receive your feedback. Thanks a lot!

**Comment by zorgiepoo** [comment, marked as answer]
[edit: this is not a common solution. See below answer for what most apps do and what we recommend them to do]

I am assuming the app you are updating still belongs to you, just that you want to update it from another process that also belongs to you, if the app is not running (this "belongs to you" point is important).

Using [sparkle-cli](https://sparkle-project.org/documentation/sparkle-cli/), is what you want similar to:
```
# This lets sparkle schedule update checks automatically (provided it has been granted permission). You can add --check-immediately if you don't want this and want to force a check.
./sparkle bundle-path --defer-install
```

I'm not saying you have to use sparkle-cli (which just wraps around Sparkle's framework with a minimal UI) but we can start discussion there. `--defer-install` updates the app immediately if the app isn't running otherwise it updates the app after it terminates, in which case Sparkle's tools (Autoupdate / Updater.app) are spawned and handles that.

You also need to tackle potential user authorization case when users don't have sufficient permission to install an app without authorizing (for example a standard user updating an app in `/Applications/` owned by admin). You can let Sparkle ask for authorization in such cases (denoted by `--interactive` in sparkle-cli), or have your daemon authorize beforehand and run Sparkle's updater under root (which is supported recently now in [2.2.0 betas](https://github.com/sparkle-project/Sparkle/releases) as long as you don't use AppKit and don't have any GUI).

**Reply by Davack** [reply]
Actually, this wraps up my question. Thanks a lot!

**Reply by zorgiepoo** [reply]
If you do go a route of an external updater, be sure the external updater and the bundle being updated are Apple code signed with the same identity.

**Comment by zorgiepoo** [comment]
The above answer I wrote might be overthinking it..

What most apps do for automatically installing updates is use `SUAutomaticallyUpdate` [described here](https://sparkle-project.org/documentation/customization/). When the app is running, updates are downloaded silently and installed on quit when possible. Did you already look at that?
